File: app/src/main/java/com/druger/aboutwork/presenters/CompaniesPresenter.java
package com.druger.aboutwork.presenters;

import com.druger.aboutwork.db.RealmHelper;
import com.druger.aboutwork.interfaces.view.CompaniesView;
import com.druger.aboutwork.model.realm.CompanyRealm;
import com.druger.aboutwork.rest.RestApi;
import com.druger.aboutwork.rest.models.CompanyResponse;
import com.druger.aboutwork.utils.rx.RxUtils;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import io.realm.OrderedRealmCollectionChangeListener;
import io.realm.RealmResults;
import moxy.InjectViewState;

import static io.realm.OrderedCollectionChangeSet.State.INITIAL;

/**
 * Created by druger on 01.05.2017.
 */

@InjectViewState
public class CompaniesPresenter extends BasePresenter<CompaniesView> {

    private RealmResults<CompanyRealm> companies;

    private OrderedRealmCollectionChangeListener<RealmResults<CompanyRealm>> realmCallback =
            (companies, changeSet) -> {
                if (companies.size() > 0 && changeSet.getState() == INITIAL) {
                    getViewState().showWatchedRecently();
                    getViewState().showCompaniesRealm();
                }
            };

    @Inject
    public CompaniesPresenter(RestApi restApi, RealmHelper realmHelper) {
        this.restApi = restApi;
        this.realmHelper = realmHelper;
    }

    public void getCompanies(String query, int page) {
        requestGetCompanies(query, page);
    }

    private void requestGetCompanies(String query, int page) {
        Disposable request = restApi.getCompany().getCompanies(query, page)
                .compose(RxUtils.httpSchedulers())
                .subscribe(this::successGetCompanies, this::handleError);
        unSubscribeOnDestroy(request);
    }

    private void successGetCompanies(CompanyResponse response) {
        getViewState().showProgress(false);
        getViewState().showCompanies(response.getItems(), response.getPages());
    }

    @Override
    protected void handleError(Throwable throwable) {
        super.handleError(throwable);
        getViewState().showProgress(false);
    }

    public void saveCompanyToDb(CompanyRealm company) {
        realmHelper.saveCompany(company);
    }

    public RealmResults<CompanyRealm> getCompaniesFromDb() {
        companies = realmHelper.getCompanies();
        companies.addChangeListener(realmCallback);
        return companies;
    }

    public void removeRealmListener() {
        companies.removeChangeListener(realmCallback);
    }
}

File: app/src/main/java/com/druger/aboutwork/presenters/CompaniesPresenter.kt
package com.druger.aboutwork.presenters

import com.druger.aboutwork.db.FirebaseHelper
import com.druger.aboutwork.interfaces.view.CompaniesView
import com.druger.aboutwork.model.Company
import com.druger.aboutwork.model.Review
import com.google.firebase.database.*
import moxy.InjectViewState
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by druger on 01.05.2017.
 */

@InjectViewState
class CompaniesPresenter : BasePresenter<CompaniesView>() {

    private lateinit var dbReference: DatabaseReference
    private var reviewEventListener: ValueEventListener? = null
    private var companyEventListener: ValueEventListener? = null

    private var reviews = ArrayList<Review>()

    override fun handleError(throwable: Throwable) {
        super.handleError(throwable)
        viewState.showProgress(false)
    }

    fun fetchReviews() {
        dbReference = FirebaseDatabase.getInstance().reference
        viewState.showProgress(true)
        reviews.clear()
        getReviews()
    }

    private fun getReviews() {
        val reviewsQuery = FirebaseHelper.getLastReviews(dbReference)
        reviewEventListener = object : ValueEventListener {
            override fun onCancelled(databaseError: DatabaseError) {
                viewState.showProgress(false)
                Timber.e(databaseError.toException())
            }

            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (snapshot in dataSnapshot.children) {
                        val review = snapshot.getValue(Review::class.java)
                        review?.let {
                            it.firebaseKey = snapshot.key
                            getCompanies(it)
                        }
                    }
                    viewState.showProgress(false)
                    reviews.reverse()
                    viewState.showReviews(reviews)
                } else {
                    viewState.showProgress(false)
                    viewState.showEmptyReviews()
                }
            }
        }
        reviewsQuery.addValueEventListener(reviewEventListener as ValueEventListener)
    }

    private fun getCompanies(review: Review?) {
        review?.companyId?.let { companyId ->
            val queryCompanies = FirebaseHelper.getCompany(dbReference, companyId)
            companyEventListener = object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.exists()) {
                        val company = dataSnapshot.getValue(Company::class.java)
                        review.name = company?.name
                        viewState.updateAdapter()
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Timber.e(databaseError.message)
                    viewState.showProgress(false)
                }
            }
            queryCompanies.addValueEventListener(companyEventListener as ValueEventListener)
            reviews.add(review)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        reviewEventListener?.let { dbReference.removeEventListener(it) }
        companyEventListener?.let { dbReference.removeEventListener(it) }
    }
}
